jobs: Move common progress tracking to new column

dt opened the pull request:
This moves all of common progress tracking -- completion percentage and
modified time -- to the new, separate progress column.

On its own this does solve progress-related writing issues as per-job
fields, such as which ranges have been exported or low/high watermarks,
are sitll in the job detail and thus require re-writing the payload to
update as the job makes progress.

Follow-up changes will need to move those fields from the Payload Detail
field to a similar job-specific detail field of the Progress column, but
this lays the groundwork for those changes.

Release note: none.

dt committed:
jobs: Move common progress tracking to new column

This moves all of common progress tracking -- completion percentage and
modified time -- to the new, separate progress column.

On its own this does solve progress-related writing issues as per-job
fields, such as which ranges have been exported or low/high watermarks,
are sitll in the job detail and thus require re-writing the payload to
update as the job makes progress.

Follow-up changes will need to move those fields from the Payload Detail
field to a similar job-specific detail field of the Progress column, but
this lays the groundwork for those changes.

Release note: none.

dt committed:
sqlmigrations: remove DisableMigrations knob

The fact that migrations in some release can be assumed to have run allows making
reasonable assumptions elsewhere, such as about what system tables exist or what
their schemas are, what formats can be expected, etc. If migrations can be skipped
those assumptions no longer hold, and the number of potential states one needs to
handle, and prove correct, grows rapidly.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26313)
<!-- Reviewable:end -->

benesch commented:
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/jobs/registry_external_test.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvmmiy9113Q_p2EUi5:-LDvmmiy9113Q_p2EUi6:b-kwapze) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sql/jobs/registry_external_test.go#L64)):*
> ```Go
> 	if e, a := storedJob, retrievedJob; !reflect.DeepEqual(e, a) {
> 		//diff := strings.Join(pretty.Diff(e, a), "\n")
> 		t.Fatalf("stored job did not match retrieved job:\n%+v\n%+v", e, a)
> ```

Debugging?

---

*[pkg/sqlmigrations/migrations.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvn2N7CKwyH96zpSJG:-LDvn2N7CKwyH96zpSJH:b-tjvok4) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sqlmigrations/migrations.go#L188)):*
> ```Go
> 	},
> 	{
> 		// Introduced in v2.0. Baked into v2.1.
> ```

```
// Introduced in v2.1.
//
// TODO(dt): Bake into v2.2.
```

---

*[pkg/sqlmigrations/migrations.go, line 807 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvnSZ14DKSQHb2f-IB:-LDvnSZ14DKSQHb2f-IC:bh5kreq) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sqlmigrations/migrations.go#L807)):*
> ```Go
> 		return false, nil
> 	}
> 	return upgradeDescsWithFn(ctx, r, tablefn, nil)
> ```

It's pretty silly to be O(n) in the number of tables in the cluster, including user tables, when all you want to do is upgrade the jobs table. I'd much prefer that you manually load/rewrite just the one table. Also this deserves a big scary comment about why you can't just `ALTER TABLE system.jobs`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sqlmigrations/migrations.go:
```diff
@@ -184,6 +184,11 @@ var backwardCompatibleMigrations = []migrationDescriptor{
 		newDescriptorIDs: databaseIDs(sessiondata.DefaultDatabaseName, sessiondata.PgDatabaseName),
 	},
+	{
+		// Introduced in v2.0. Baked into v2.1.
```
What does introduced in v2.0 mean? I'm confused.

mjibson commented on pkg/sql/jobs/jobs.proto:
```diff
@@ -130,12 +130,12 @@ message Payload {
   int64 started_micros = 3;
   int64 finished_micros = 4;
-  int64 modified_micros = 5;
+  int64 legacy_modified_micros = 5;
```
why do these need to be kept around instead of made reserved?

madelynnblue reviewed (commented):
Overall LGTM. I wonder if we should wait to merge this until we do the work to fix IMPORT to use this too to avoid another migration? Or do you expect it to be fine?

dt committed:
jobs: Move job-specific progress info to Progress

This moves the job-specific progress information, like what ranges have
been imported or low/high watermarks, to a sub-field of the `progress`
column and its corresponding Progress proto.

By definition, progress info changes as the job makes progress, so it
needs to be re-written repeatedly, while much of the job's details, for
example input file URIs, do not change at all. When these are subfields
of the same Details proto, changing the progress information requires
rewriting the entire column (and indeed the whole column family). If the
details are large, the MVCC history created by these rewrites can add up
quickly, causing problems as the row grows too large.

This change should mean that only the very small progress field needs to
be re-written regularly, while the details field should only require
being re-written when a fixed number of times -- when started or stopped
or at major milestones to record large fields that would make the small
progress field too large (like the sampled data for in IMPORT job).

Release note: none.

dt commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/jobs/jobs.proto, line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb0hURw-jHCCJgx-r2-133:-LEAPJTz-fh3Hs_gJiwu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sql/jobs/jobs.proto#L133)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

why do these need to be kept around instead of made reserved?
</blockquote></details>

Done.

---

*[pkg/sql/jobs/registry_external_test.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvmmiy9113Q_p2EUi5:-LEAMKFHEbIoJKodKgBV:b-wc9vok) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sql/jobs/registry_external_test.go#L64)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Debugging?
</blockquote></details>

oh, yeah, interestingly, `pretty.Diff` seemed to be stuck in an infinite loop.

---

*[pkg/sqlmigrations/migrations.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvn2N7CKwyH96zpSJG:-LEAMf0C0IZEfnahDs5H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sqlmigrations/migrations.go#L188)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

```
// Introduced in v2.1.
//
// TODO(dt): Bake into v2.2.
```
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 188 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb-caBCO8x7Toud-r2-188:-LEAMgOcD8D_Uz9kYI1G:b-tunaa6) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sqlmigrations/migrations.go#L188)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

What does introduced in v2.0 mean? I'm confused.
</blockquote></details>

typo, meant 2.1. it's when we added the migration, so that we know when we can remove it.

---

*[pkg/sqlmigrations/migrations.go, line 807 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvnSZ14DKSQHb2f-IB:-LEAP58W8Cpuof72-Mb3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/008a07094fb2dacaf164f0ae13829b6e1a86ea60/pkg/sqlmigrations/migrations.go#L807)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

It's pretty silly to be O(n) in the number of tables in the cluster, including user tables, when all you want to do is upgrade the jobs table. I'd much prefer that you manually load/rewrite just the one table. Also this deserves a big scary comment about why you can't just `ALTER TABLE system.jobs`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

dt commented:
I wanted to get the first, more mechanical part reviewed on its own, but I've now added the migration of the job-specific progress info to the new field (though job.update still rewrites everything for now).

I want to get the job control tests unskipped ASAP, but right now one of them hangs if I do that. I'll probably look at doing that in a followup though since this is already becoming a large change.

benesch commented:
Migration stuff LGTM! Leaving the rest of the review to other folks.

---

Review status: 12 of 29 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sqlmigrations/migrations.go, line 817 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LEAYDKBD0uZXhujgR_g:-LEAYDKBD0uZXhujgR_h:b-hh9nu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdee5e7cda01e364eeaa3feb8d388bd63e6e0820/pkg/sqlmigrations/migrations.go#L817)):*
> ```Go
> 			return err
> 		}
> 		desc.UpVersion = true
> ```

If you wind up going back through CI, either copy the big scary comment about why we can set `desc.UpVersion` directly or call `SetUpVersion`. Or ignore if UpVersion is going away sometime soon; I think I saw an issue about that.

---

*[pkg/sqlmigrations/migrations.go, line 826 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LEAYh1D2t8lkms0Ie8x:-LEAYh1D2t8lkms0Ie8y:bwgbbez) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdee5e7cda01e364eeaa3feb8d388bd63e6e0820/pkg/sqlmigrations/migrations.go#L826)):*
> ```Go
> 			return err
> 		}
> 		return txn.Run(ctx, b)
> ```

optional nit: use txn.Put directly:

```
return txn.Put(sqlbase.MakeDescMetadataKey(desc.ID), sqlbase.WrapDescriptor(desc))
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-:-LEAZ1Yx1Cixg4-HA_y9:bk2ru3x)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 10 of 29 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sqlmigrations/migrations.go, line 817 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LEAYDKBD0uZXhujgR_g:-LEAjsaU1TqTfOw2xA4R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdee5e7cda01e364eeaa3feb8d388bd63e6e0820/pkg/sqlmigrations/migrations.go#L817)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

If you wind up going back through CI, either copy the big scary comment about why we can set `desc.UpVersion` directly or call `SetUpVersion`. Or ignore if UpVersion is going away sometime soon; I think I saw an issue about that.
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 826 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LEAYh1D2t8lkms0Ie8x:-LEAjsFJ1I6g2EZ6lfC-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdee5e7cda01e364eeaa3feb8d388bd63e6e0820/pkg/sqlmigrations/migrations.go#L826)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

optional nit: use txn.Put directly:

```
return txn.Put(sqlbase.MakeDescMetadataKey(desc.ID), sqlbase.WrapDescriptor(desc))
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sql/jobs/jobs.proto:
```diff
@@ -130,12 +130,12 @@ message Payload {
     (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sqlbase.ID"
   ];
-  float fraction_completed = 7;
+  float legacy_fraction_completed = 7;
```
does this also need to be marked reserved?

mjibson commented on pkg/sqlmigrations/migrations.go:
```diff
@@ -780,3 +786,40 @@ func createDefaultDbs(ctx context.Context, r runner) error {
 	return err
 }
+
+func addJobsProgress(ctx context.Context, r runner) error {
```
is there a test we can add somewhere to ensure that this migration has run? like a test that expects a specific output of `show create table system.jobs`?

mjibson commented on pkg/sql/jobs/jobs.go:
```diff
@@ -160,27 +176,49 @@ func (j *Job) Started(ctx context.Context) error {
 // Progressed updates the progress of the tracked job. It sets the job's
 // FractionCompleted field to the value returned by progressedFn and persists
-// progressedFn's modifications to the job's details, if any.
+// progressedFn's modifications to the job's pgoress details, if any.
```
pgoress

mjibson commented on pkg/sql/jobs/jobs.proto:
```diff
@@ -92,14 +99,21 @@ message ImportDetails {
   repeated Table tables = 1 [(gogoproto.nullable) = false];
 }
 
+message ImportProgress {
```
Yay. This is much better than the per-table progress before.

dt commented:
Review status: 4 of 29 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/jobs/jobs.go, line 195 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb0hURw-jHCCJgv-r5-195:-LEBGmfvBe6lFEOOW3CH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68181c2d775d6e77e28fb5fe21dfb2581b478eec/pkg/sql/jobs/jobs.go#L195)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

pgoress
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.proto, line 138 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb0hURw-jHCCJgx-r4-138:-LEBHP3V2mWz85G3OLDA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12bd0a91db4ba4bb70334e6fc8a7ab9dd4b008ee/pkg/sql/jobs/jobs.proto#L138)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

does this also need to be marked reserved?
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.proto, line 107 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb0hURw-jHCCJgx-r5-107:-LEBGn2W9wh2ErzZ9Zud:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68181c2d775d6e77e28fb5fe21dfb2581b478eec/pkg/sql/jobs/jobs.proto#L107)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Yay. This is much better than the per-table progress before.
</blockquote></details>

Done.

---

*[pkg/sqlmigrations/migrations.go, line 790 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb-caBCO8x7Toud-r4-790:-LEBHAuY-On22MTeS8fo:bx1txfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/12bd0a91db4ba4bb70334e6fc8a7ab9dd4b008ee/pkg/sqlmigrations/migrations.go#L790)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

is there a test we can add somewhere to ensure that this migration has run? like a test that expects a specific output of `show create table system.jobs`?
</blockquote></details>

all the jobs code mentions the column, so the pkg/sql/jobs tests fail immediately if the migration isn't run.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 4 of 29 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sqlmigrations/migrations.go, line 790 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb-caBCO8x7Toud-r4-790:-LEBQOwR9LtqlSoPfDn0:b-n13j66) ([raw file](https://github.com/cockroachdb/cockroach/blob/12bd0a91db4ba4bb70334e6fc8a7ab9dd4b008ee/pkg/sqlmigrations/migrations.go#L790)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

all the jobs code mentions the column, so the pkg/sql/jobs tests fail immediately if the migration isn't run.
</blockquote></details>

I meant to assert existence of the column family, not the column.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 4 of 29 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sqlmigrations/migrations.go, line 790 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb-caBCO8x7Toud-r4-790:-LEBfEPXBQ0lkvmYIucm:bv9m06p) ([raw file](https://github.com/cockroachdb/cockroach/blob/12bd0a91db4ba4bb70334e6fc8a7ab9dd4b008ee/pkg/sqlmigrations/migrations.go#L790)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I meant to assert existence of the column family, not the column.
</blockquote></details>

Ah. that has some coverage in the show kv trace, since it switched from 1 put to 2 puts, but i'll add an explicit `show create table` test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 4 of 29 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sqlmigrations/migrations.go, line 790 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26313#-LDvQkb-caBCO8x7Toud-r4-790:-LEBi5dG4dSPuL9WiHcE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12bd0a91db4ba4bb70334e6fc8a7ab9dd4b008ee/pkg/sqlmigrations/migrations.go#L790)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Ah. that has some coverage in the show kv trace, since it switched from 1 put to 2 puts, but i'll add an explicit `show create table` test.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26313)*
<!-- Sent from Reviewable.io -->

dt commented:
bors r+
